server/src/testFixtures/java/io/crate/types/DataTypeTestCase.java

```diff
@@ -147,20 +164,43 @@ protected void doReferenceResolveTest(DataType<T> type, String definition, T dat
             docValueImpl.startCollect(collectorContext);
             docValueImpl.setNextReader(readerContext);
             docValueImpl.setNextDocId(nextDoc);
-            assertEquals((T) docValueImpl.value(), data);
+
+            var value = docValueImpl.value();
+            if (value instanceof List<?> l) {
+                assertListEquals((List<T>) l, (List<T>) data);
+            } else {
+                assertEquals((T) docValueImpl.value(), (T) data);
+            }
         }
     }
 
     protected void assertEquals(T actual, T expected) {
         assertThat(actual).isEqualTo(expected);
     }
 
+    protected void assertListEquals(List<T> actual, List<T> expected) {
+        assertThat(actual).hasSize(expected.size());
+        for (int i = 0; i < actual.size(); i++) {
+            assertEquals(actual.get(i), expected.get(i));
+        }
+    }
+
+    private static String toTypeSignature(DataType<?> dataType) {
+        if (dataType instanceof ObjectType o) {
+            assert o.innerTypes().size() == 1;
+            return "object as (x " + o.innerType("x").getTypeSignature().toString() + ")";
+        } else {
+            return dataType.getTypeSignature().toString();
+        }
+    }
```

Comment:
Is this really needed? `ObjectType.getTypeSignature()` should take care of inner types.

Reply:
Unfortunately yes - `ObjectType.getTypeSignature().toString` here returns `object(text,"x" byte)` when what we want is `object as (x byte)`.